src/librustc_parse/parser/pat.rs

```diff
@@ -459,16 +459,27 @@ impl<'a> Parser<'a> {
     /// Parse `&pat` / `&mut pat`.
     fn parse_pat_deref(&mut self, expected: Expected) -> PResult<'a, PatKind> {
         self.expect_and()?;
+        self.recover_lifetime_in_deref_pat();
         let mutbl = self.parse_mutability();
+        let subpat = self.parse_pat_with_range_pat(false, expected)?;
+        Ok(PatKind::Ref(subpat, mutbl))
+    }
 
+    fn recover_lifetime_in_deref_pat(&mut self) {
         if let token::Lifetime(name) = self.token.kind {
-            let mut err = self.fatal(&format!("unexpected lifetime `{}` in pattern", name));
-            err.span_label(self.token.span, "unexpected lifetime");
-            return Err(err);
-        }
+            self.bump(); // `'a`
 
-        let subpat = self.parse_pat_with_range_pat(false, expected)?;
-        Ok(PatKind::Ref(subpat, mutbl))
+            let span = self.prev_span;
+            self.struct_span_err(span, &format!("unexpected lifetime `{}` in pattern", name))
+                .span_label(span, "unexpected lifetime")
```

Comment:
Having the label the message and the suggestion all point at the same span is too verbose. Remove the `span_label` so the output is

```

error: unexpected lifetime `'a` in pattern
  --> $DIR/lifetime-in-pattern-recover.rs:2:10
   |
LL |     let &'a x = &0;
   |          ^^ help: remove the lifetime
```